### ISSUE TITLE
editing any email template - errors shown before and after save 

### ISSUE THREAD

**Issue by dallenk** [iomad/iomad#1933]
PHP 8.0.30 (cli) (built: Sep  2 2023 08:05:13) ( NTS )
commit 1caf6bfcc793235de00afba7868715b2fe2f7517 (HEAD -> IOMAD_402_STABLE, tag: IOMAD_402_06, origin/IOMAD_402_STABLE)

Warning: Undefined property: stdClass::$repeatday in /var/www/vhosts/sandbox.domain.ca/iomad/local/email/classes/forms/template_edit_form.php on line 203

**Comment by turf212** [comment]
This is part of the database record from mdl_email_templateset_templates.  It should exist as a field there so should never not be set.  Can you please check your DB and that table?

**Comment by dallenk** [comment]
is this helpful at all?

```
MariaDB [db_sandbox_firetrain_ca]> select id,templateset,name from mdl_email_templateset_template
+----+-------------+-----------------------------------+
| id | templateset | name                              |
+----+-------------+-----------------------------------+
|  1 |           1 | admin_deleted                     |
|  2 |           1 | advertise_classroom_based_course  |
|  3 |           1 | approval                          |
|  4 |           1 | company_licenseassigned           |
|  5 |           1 | company_suspended                 |
|  6 |           1 | company_unsuspended               |
|  7 |           1 | completion_course_supervisor      |
|  8 |           1 | completion_course_user            |
|  9 |           1 | completion_digest_manager         |
| 10 |           1 | completion_expiry_warn_supervisor |
| 11 |           1 | completion_warn_manager           |
| 12 |           1 | completion_warn_supervisor        |
| 13 |           1 | completion_warn_user              |
| 14 |           1 | course_classroom_approval         |
| 15 |           1 | course_classroom_approval_request |
| 16 |           1 | course_classroom_approved         |
| 17 |           1 | course_classroom_denied           |
| 18 |           1 | course_classroom_manager_denied   |
| 19 |           1 | course_completed_manager          |
| 20 |           1 | course_not_started_warning        |
| 21 |           1 | expire                            |
| 22 |           1 | expire_manager                    |
| 23 |           1 | expiry_warn_manager               |
| 24 |           1 | expiry_warn_user                  |
| 25 |           1 | invoice_ordercomplete             |
| 26 |           1 | invoice_ordercomplete_admin       |
| 27 |           1 | license_allocated                 |
| 28 |           1 | license_reminder                  |
| 29 |           1 | license_removed                   |
| 30 |           1 | licensepoolexpiring               |
| 31 |           1 | licensepoolwarning                |
| 32 |           1 | microlearning_nugget_reminder1    |
| 33 |           1 | microlearning_nugget_reminder2    |
| 34 |           1 | microlearning_nugget_scheduled    |
| 35 |           1 | password_update                   |
| 36 |           1 | user_added_to_course              |
| 37 |           1 | user_create                       |
| 38 |           1 | user_deleted                      |
| 39 |           1 | user_programcompleted             |
| 40 |           1 | user_promoted                     |
| 41 |           1 | user_removed_from_event           |
| 42 |           1 | user_reset                        |
| 43 |           1 | user_signed_up_for_event          |
| 44 |           1 | user_suspended                    |
| 45 |           1 | user_unsuspended                  |
| 46 |           2 | admin_deleted                     |
| 47 |           2 | advertise_classroom_based_course  |
| 48 |           2 | approval                          |
| 49 |           2 | company_licenseassigned           |
| 50 |           2 | company_suspended                 |
| 51 |           2 | company_unsuspended               |
| 52 |           2 | completion_course_supervisor      |
| 53 |           2 | completion_course_user            |
| 54 |           2 | completion_digest_manager         |
| 55 |           2 | completion_expiry_warn_supervisor |
| 56 |           2 | completion_warn_manager           |
| 57 |           2 | completion_warn_supervisor        |
| 58 |           2 | completion_warn_user              |
| 59 |           2 | course_classroom_approval         |
| 60 |           2 | course_classroom_approval_request |
| 61 |           2 | course_classroom_approved         |
| 62 |           2 | course_classroom_denied           |
| 63 |           2 | course_classroom_manager_denied   |
| 64 |           2 | course_completed_manager          |
| 65 |           2 | course_not_started_warning        |
| 66 |           2 | expire                            |
| 67 |           2 | expire_manager                    |
| 68 |           2 | expiry_warn_manager               |
| 69 |           2 | expiry_warn_user                  |
| 70 |           2 | invoice_ordercomplete             |
| 71 |           2 | invoice_ordercomplete_admin       |
| 72 |           2 | license_allocated                 |
| 73 |           2 | license_reminder                  |
| 74 |           2 | license_removed                   |
| 75 |           2 | licensepoolexpiring               |
| 76 |           2 | licensepoolwarning                |
| 77 |           2 | microlearning_nugget_reminder1    |
| 78 |           2 | microlearning_nugget_reminder2    |
| 79 |           2 | microlearning_nugget_scheduled    |
| 80 |           2 | password_update                   |
| 81 |           2 | user_added_to_course              |
| 82 |           2 | user_create                       |
| 83 |           2 | user_deleted                      |
| 84 |           2 | user_programcompleted             |
| 85 |           2 | user_promoted                     |
| 86 |           2 | user_removed_from_event           |
| 87 |           2 | user_reset                        |
| 88 |           2 | user_signed_up_for_event          |
| 89 |           2 | user_suspended                    |
| 90 |           2 | user_unsuspended                  |
| 91 |           1 | trainingevent_not_selected        |
| 92 |           1 | user_removed_from_event_teacher   |
| 93 |           1 | user_removed_from_event_waitlist  |
| 94 |           2 | trainingevent_not_selected        |
| 95 |           2 | user_removed_from_event_teacher   |
| 96 |           2 | user_removed_from_event_waitlist  |
+----+-------------+-----------------------------------+
96 rows in set (0.018 sec)

```

**Comment by turf212** [comment]
try

`select * from mdl_email_templateset_template where id = 1 \G`

Note the \G bit is instead of ; and will display it in a single column instead of a row.

**Comment by dallenk** [comment]
```
MariaDB [db_sandbox_firetrain_ca]> select * from mdl_email_templateset_templates where id = 1 \G
*************************** 1. row ***************************
                id: 1
       templateset: 1
              name: admin_deleted
              lang: en
           subject: Account has been demoted
              body: <p>Hello {User_Firstname} {User_Lastname},</p>
<p>Your admin privileges have been revoked on {SiteURL}. You no longer have access to manage your company</p>
<p>Your email: {User_Email}</p><p>If you feel that this is in error, please contact your manager.</p>
         signature: NULL
          disabled: 1
   disabledmanager: 1
disabledsupervisor: 1
           emailto: NULL
      emailtoother: NULL
           emailcc: NULL
      emailccother: NULL
         emailfrom: NULL
    emailfromother: NULL
      emailreplyto: NULL
 emailreplytoother: NULL
      repeatperiod: 0
       repeatvalue: 0
        repeateday: 0
emailfromothername: NULL
1 row in set (0.043 sec)

```

**Comment by turf212** [comment]
Column name doesn't match the object element.

Column is repeateday and it's looking for repeatday.

Both my 4.1 and 4.2 tables look like this:

mysql> desc mdl_email_templateset_templates;
+--------------------+---------------+------+-----+---------+----------------+
| Field              | Type          | Null | Key | Default | Extra          |
+--------------------+---------------+------+-----+---------+----------------+
| id                 | bigint        | NO   | PRI | NULL    | auto_increment |
| templateset        | bigint        | NO   |     | NULL    |                |
| name               | varchar(50)   | NO   |     |         |                |
| lang               | varchar(20)   | NO   |     |         |                |
| subject            | varchar(255)  | YES  |     | NULL    |                |
| body               | longtext      | YES  |     | NULL    |                |
| signature          | longtext      | YES  |     | NULL    |                |
| disabled           | tinyint(1)    | NO   |     | 0       |                |
| disabledmanager    | tinyint(1)    | NO   |     | 0       |                |
| disabledsupervisor | tinyint(1)    | NO   |     | 0       |                |
| emailto            | varchar(1333) | YES  |     | NULL    |                |
| emailtoother       | varchar(1333) | YES  |     | NULL    |                |
| emailcc            | varchar(1333) | YES  |     | NULL    |                |
| emailccother       | varchar(1333) | YES  |     | NULL    |                |
| emailfrom          | varchar(1333) | YES  |     | NULL    |                |
| emailfromother     | varchar(1333) | YES  |     | NULL    |                |
| emailreplyto       | varchar(1333) | YES  |     | NULL    |                |
| emailreplytoother  | varchar(1333) | YES  |     | NULL    |                |
| repeatperiod       | bigint        | NO   |     | 0       |                |
| repeatvalue        | bigint        | NO   |     | 0       |                |
| repeatday          | bigint        | NO   |     | 0       |                |
| emailfromothername | varchar(1333) | YES  |     | NULL    |                |
+--------------------+---------------+------+-----+---------+----------------+


It was renamed in version 2017080707 of local/email
    if ($oldversion < 2017080707) {

        // Rename field repeateday on table email_template to repeatday.
        $table = new xmldb_table('email_template');
        $field = new xmldb_field('repeateday', XMLDB_TYPE_INTEGER, '10', null, XMLDB_NOTNULL, null, '0', 'repeatvalue');

        // Launch rename field repeatday.
        $dbman->rename_field($table, $field, 'repeatday');

        // Email savepoint reached.
        upgrade_plugin_savepoint(true, 2017080707, 'local', 'email');
    }

But it looks like the value wasn't changed in the install.xml file till commit 7a3d47f49320 so they could be out of step

I've updated the code and version of local email to fix in commit 6d96e793a851d33a755f6b05141bb97c823b6d24
